lib/ansible/modules/packaging/os/snap.py

```diff
@@ -0,0 +1,193 @@
+    classic:
+        description:
+            - Confinement policy. The classic confinment allows a snap to have
+              the same level of access to the system as "classic" packages,
+              like those managed by APT. This option corresponds to the --classic argument.
+        type: bool
+        required: false
+        default: False
+
+author:
+    - Victor Carceler (vcarceler@iespuigcastellar.xeill.net)
+    - Stanislas Lange (angristan) <angristan@pm.me>
+'''
+
+EXAMPLES = '''
+# Install "foo" snap
+- name: Install foo
+  snap:
+    name: foo
+
+# Remove "foo" snap
+- name: Remove foo
+  snap:
+    name: foo
+    state: absent
+
+# Install a snap with classic confinement
+- name: Install "foo" with option --classic
+  snap:
+    name: foo
+    classic: yes
+'''
+
+RETURN = '''
+classic:
+    description: Whether or not the snaps were installed with the classic confinement
+    type: boolean
+    returned: When snaps are installed
+cmd:
+    description: The command that was executed on the host
+    type: string
+    returned: When changed is true
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+import re
+
+
+def snap_exists(module, snap_name):
+    snap_path = module.get_bin_path("snap", True)
+    cmd_parts = [snap_path, 'info', snap_name]
+    cmd = ' '.join(cmd_parts)
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    return rc == 0
+
+
+def is_snap_installed(module, snap_name):
+    snap_path = module.get_bin_path("snap", True)
+    cmd_parts = [snap_path, 'list', snap_name]
+    cmd = ' '.join(cmd_parts)
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    return rc == 0
+
+
+def install_snaps(module, snap_names):
+    snaps_not_installed = list()
+    for snap_name in snap_names:
+        if is_snap_installed(module, snap_name):
+            continue
+        snaps_not_installed.append(snap_name)
+
+    if not snaps_not_installed:
+        module.exit_json(classic=module.params['classic'], changed=False)
+
+    if module.check_mode:
+        module.exit_json(classic=module.params['classic'], changed=True)
+
+    classic = '--classic' if module.params['classic'] else ''
+
+    snap_path = module.get_bin_path("snap", True)
+    cmd_parts = [snap_path, 'install'] + snaps_not_installed + [classic]
+    cmd = ' '.join(cmd_parts)
+
+    # Actually install the snaps
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    if rc == 0:
+        module.exit_json(classic=module.params['classic'], changed=True, cmd=cmd, stdout=out, stderr=err)
+    else:
+        m = re.match(r'^error: This revision of snap "(?P<package_name>\w+)" was published using classic confinement', err)
+        if m is not None:
+            err_pkg = m.group('package_name')
+            msg = "Couldn't install {name} because it requires classic confinement".format(name=err_pkg)
+        else:
+            # The error is not related to the confinement
+            msg = "Something went wrong"
```

Comment:
Also, error message is not very useful. It would be nice to reveal some details.

Reply:
That's why we have `stderr`

Reply:
Not exactly. Of course stderr helps. But try to compare, from the UX point of view, "Ooops! Snap installation failed, please examine logs and error output for more details." vs. "Error happened." vs. "Some error here." vs. "Something unexpected happened."

Reply:
What do you prefer to see as a user? "This module doesn't work" or "This action failed, but here's some helpful hints"?

Reply:
I totally understand :)
But how am I supposed to know what could go wrong? The code you added for when the snap requires classic is great. But we only added it it because I stumbled upon the error.
The best I can do is "Something when wrong while executing `cmd`". I cannot parse imaginary error messages 🤔

Reply:
Yes, that's better.
You cannot parse some messages, but *you know* in which part of *your* program "something went wrong" and that's what you need to report to the user. I'd do:
```
Ooops! Snap installation failed while executing '{cmd}', please examine logs and error output for more details."
```

Reply:
OK 👍https://github.com/ansible/ansible/pull/44939/commits/8d56e9a27e7a5688baa1eea49657b0a3a9844348